### ISSUE TITLE
[Bug] [Android] CollectionView jittering on Android with with GridItemsLayout

### ISSUE THREAD

**Issue by yurkinh** [xamarin/Xamarin.Forms#8718]
### Description
I tried to reproduce UI of one popular music app (DI.FM).
During implementation of page with styles I have encountered issue with collectionview and images. Fast scroll on CollectionView with lots of images produces jittering on Android.
I have tried Image, ffImageLoading:CachedImage (a bit better), Image with  GlideX(a bit better).
While it works perfectly smooth on ios even with ordinary Image control. Also I have tried different sizing strategies.

### Steps to Reproduce

1. Start reproduction project on Android and open RadioChannels "Popular" tab or just "Popular" page from side menu. (Same on ios for comparison)

### Expected Behavior
No jittering, smooth scrolling

### Actual Behavior
Jittering

### Basic Information

- Version with issue:
- Nuget Packages: Xamarin.Forms 4.3.0.991211 
- Affected Devices: Android devices and emulators

### Reproduction Link

[DIFM.zip](https://github.com/xamarin/Xamarin.Forms/files/3905802/DIFM.zip)

**Comment by jfversluis** [comment]
I don't see anything weird with how the images are loaded. Images are not too big. Yet the scrolling is not very smooth.

**Comment by yurkinh** [comment]
Yes. But on ios scrolling is very smooth. 
Also you can download original [app] (https://play.google.com/store/apps/details?id=com.audioaddict.di&hl=en) from PlayMarket  and check the same page. Scrolling is very smooth on Android in original app with approximately same amount of images. Original app is written with (Xamarin.Native)

**Comment by Alex-111** [comment]
Same problem here...
Scrolling jitters a lot on Android...
Any news for this issue?

**Comment by yurkinh** [comment]
@Alex-111  Tried latest stable version and still the same. But a preview of 4.6.0 looks a bit better

**Comment by Alex-111** [comment]
Also just tried the preview of 4.6. Unfortunately not much change here….

**Comment by rudyspano** [comment]
Hello. Same problem on my side... Even with LinearItemsLayout
I've tried: FFIL (better) and GlideX (really better).
In the output console, we can see many garbage collections even on a simple projet.

After analysis with Xamarin Profiler, it seems there's a lot of BindableProperty and linked mechanisms created in memory when scrolling (BindablePropertyContext, ...). I guess, that's a consequence of recycling...

When i don't use DataBinding, things are better but not perfects... Not relying on DataBinding specially with images is advised by the FFIL author for example....
You should try even if it's not cool...
 
I've also seen the same kind of problems with ListView...

If someone has some best practices/solutions on this topic, it will be really appreciated :)

**Comment by Alex-111** [comment]
I can confirm this. Lots of garbage collections in the debug output… But not sure if this alone is the reason for the heavy jitter..
Unfortunately, it is event not smooth on high end smartphones like Samsung S20, which is not accepted by customers...

**Comment by JMNewsome** [comment]
Any progress on this issue? This is causing serious issues within our project and we can't release due to this issue

**Comment by yurkinh** [comment]
Latest version in combination with GlideX.Forms looks pretty good but still not perfect as iOS version of CollectionView

**Comment by Happypig375** [comment]
@JMNewsome Don't expect the Xamarin.Forms to be quick on resolving issues...

**Comment by MSiccDev** [comment]
Any news on this? This one makes developing on Android really painful....

**Comment by brminnick** [comment]
I'm seeing the same behavior in my GitTrends app.

Here's a video one of the users posted on Twitter running GitTrends v1.2.0:
https://twitter.com/anaseeen/status/1281610967336595458?s=20
- 4GB RAM
- Media Tek Helio octa core P23
- Android 9

GitTrends v1.2.0 Source Code:
https://github.com/brminnick/GitTrends/releases/tag/v1.2.0

**Comment by brminnick** [comment]
After a quick dig into this, it may be related to garbage collection.

When scrolling, I noticed that the UI pauses each time the garbage collector executes. This can be seen in the application output logs: 
> Explicit concurrent copying GC freed 19623(1217KB) AllocSpace objects, 0(0B) LOS objects, 43% free, 31MB/55MB, paused 53us total 63.916ms

**Comment by JMNewsome** [comment]
Yeah, this is a very big performance problem, can we get some kind of higher priority on this item?

**Comment by brminnick** [comment]
# Workaround

I found that two things were causing the UI to jitter when the user scrolls a CollectionView:
- Bindings
- Garbage Collection

I was able to solve this by removing bindings from my `DataTemplate` and increase the size of the Android Nursery to decrease the number of garbage collections.

If you like to see exactly how I solved it, here is the Pull Request that fixed it in my GitTrends app: https://github.com/brminnick/GitTrends/pull/143

| Before (Substantial Scrolling Jitter) | After (No Scrolling Jitter) | 
| -------------------------- | --------------- |
| ![BeforeFix](https://user-images.githubusercontent.com/13558917/87251186-9a005f00-c41e-11ea-90e7-d3ac9d58e56d.gif) | ![AfterFix](https://user-images.githubusercontent.com/13558917/87251189-a5ec2100-c41e-11ea-86f8-089a9d1ce616.gif) |

## Removing Bindings

To remove bindings, we can use a [`DataTemplateSelector`](https://docs.microsoft.com/xamarin/xamarin-forms/app-fundamentals/templates/data-templates/selector?WT.mc_id=xamarinforms-github-bramin) to pass the `BindingContext` directly into our `DataTemplate` and assign the values on initialization.

### Before Removing Bindings
```csharp
using Xamarin.Forms.Markup;

class CollectionViewPage : ContentPage
{
    Content = new CollectionView
    {
        ItemTemplate = new ImageDataTemplate()
    }.Bind(CollectionView.ItemsSourceProperty, nameof(CollectionViewModel.ImageList));
}

class ImageModel
{
    public string ImageTitle { get; set; }
    public string ImageUrl { get; set; }
}

class ImageDataTemplate : DataTemplate
{
    public MyDataTemplate() : base(() => CreateDataTemplate())
    {

    }

    static View CreateDataTemplate() => new StackLayout
    {
        Children = 
        {
            new Image().Bind(Image.SourceProperty, nameof(ImageModel.ImageUrl))
            new Label().Bind(Label.TextProperty, nameof(ImageModel.ImageTitle))
        }
   }
}
```

### After Removing Bindings
```csharp
using Xamarin.Forms.Markup;

class CollectionViewPage : ContentPage
{
    Content = new CollectionView
    {
        ItemTemplate = new ImageDataTemplateSelector()
    }.Bind(CollectionView.ItemsSourceProperty, nameof(CollectionViewModel.ImageList));
}

class ImageModel
{
    public string ImageTitle { get; set; }
    public string ImageUrl { get; set; }
}

class ImageDataTemplateSelector : DataTemplateSelector
{

    protected override DataTemplate OnSelectTemplate(object item, BindableObject container) => new  ImageDataTemplate((ImageModel)item);

    class ImageDataTemplate : DataTemplate
    {
        public MyDataTemplate(ImageModel imageModel) : base(() => CreateDataTemplate(imageModel))
        {

        }

        static View CreateDataTemplate(ImageModel imageModel) => new StackLayout
        {
            Children = 
            {
                new Image { Source = imageModel.ImageUrl },
                new Label { Text = imageModel.ImageTitle }
            }
       }
}
```

## Increate Nursery Size to Decrease Garbage Collection

We can [set the size of the Android Nursery by setting it in the `MONO_GC_PARAMS`](https://docs.microsoft.com/xamarin/android/internals/garbage-collection?WT.mc_id=xamarinforms-github-bramin#configuration):

1. In the Xamarin.Android project, create a new file called `GarbageCollector.config`
2. In `GarbageCollector.config`, enter the following code:

```bash
MONO_GC_PARAMS=nursery-size=64m
```

3. In Visual Studio, in the Solution Explorer, right-click on `GarbageCollector.config`
4. In the right-click menu, select *BuildAction* > *AndroidEnvironment*

<img width="710" alt="Screen Shot 2020-07-12 at 8 53 36 AM" src="https://user-images.githubusercontent.com/13558917/87250963-388bc080-c41d-11ea-996b-6c16cd718646.png">

**Comment by rudyspano** [comment]
Thank you @brminnick . I will try this.
Concerning the code refactoring Binding => Affectation, create a DataTemplate in code-behind seems to me a little bit unusual (especially with more complex cells). But I guess, same result could be obtained with Xaml and OnBindingContextChanged overriding. DataBinding is however powerfull and that's too bad that we cannot use it properly here :(. Thanks again

**Comment by Alex-111** [comment]
@brminnick : Thanks for suggesting some workarounds!

- regarding the nursery-size: This alone had no effect in my case - unfortunately….

- regarding the binding: In your case you do not update the view on data change. This is needed in my case. So unfortunately we cannot use your workaround...

any further suggestions are appreciated...

**Comment by MSicc** [comment]
I am in the same situation as @Alex-111 ... 
I had nursery-size and soft-heap-limit already 128M, that didn't change anything.
My DataTemplates are updating the the view as well on PropertyChanges and are a bit more complex. Binding makes those changes easy and they are working perfectly on iOS and UWP... There must be another solution/workaround...

**Comment by roubachof** [comment]
I found here: https://www.sharpnado.com/gittrends-lags/ that the GC Major collection is always growing.
In the test I made, I start from the GitTrends app repo with the CollectionView displaying empty frames.
After 1 min of scrolling back and fourth, the GC major collection grows from 4352k to 44512k

**Comment by rudyspano** [comment]
In our side: complex cells: one big image + up to 5 small images (stickers) and more or less complex controls.  
Using GlideX
-Removing all bindings (previously ~30 by cell (controls and sub controls))
-Setting MONO_GC_PARAMS=nursery-size=512m (I know that's a lot...)

We achieved to have no more scroll lags on performant devices=> No more Garbage Collections in output
We've seen an enhancement on other (older) devices and no negative impact (regarding the nursery size)

Like we observed previously, Garbage collections are caused by intensive objects creations (BindableProperty mechanisms): not sure about any memory leak...

Thanks @brminnick

**Comment by roubachof** [comment]
@rudyspano setting the nursery to 512m is like disabling GC... 
So of course you won't have gc issues...
...until you reach the 512m limit and you will experience something really weird cause a collection of a 512m nursery is not something you see every day.
It's also probable that android will kill your app before even reaching the nursery size limit.

**Comment by rudyspano** [comment]
Thanks @roubachof , I will check with a more acceptable value.

**Comment by JMNewsome** [comment]
I think this garbage collection issue is the number one factor in Xamarin Forms for Android running terribly and should be something that is prioritized.

**Comment by libin85** [comment]
It's pretty sad to see Xamarin go down this path. The listview is not recommended any more and collection view is pretty much not usable. As a developer I dont know how am I gonna use xamarin to build apps, or even support the apps that we have build over the years.

**Comment by JMNewsome** [comment]
I agree, what I love about Xamarin is the ability to use C# and shared between the platform, code execution speed for me on a compiled app is NOT the problem. The problem is the UI thread, which of course will cause everything else to appear to run bad. If the UI thread didn't feel like it was held together with duct tape and staples, It would be so much better

**Comment by Deepfreezed** [comment]
@brminnick, is this an issue when creating UI controls from code behind? I tested this out with one of my CollectionView templates. Creating UI control from code behind causes the Nursery to become full frequently, in turn calling the GC constantly. Using XAML, Nursery full warning is less frequent.

**Comment by brminnick** [comment]
@Deepfreezed This is not a XAML-specific nor a C#-specific problem.

This is a problem related to bindings and garbage collection.

**Comment by kerbooo** [comment]
Same issue is here for me. Android performance is very bad.

**Comment by ilsur-dev111** [comment]
Will the issue be resolved in the next release of Xamarin Forms? Due to this issue, we are unable to distribute the application.

**Comment by andersonvieiragomeslopes** [comment]
This problem is really very serious, I changed the listview for collectionview by official recommendation, now a problem like this arises. It is very sad to see that it has been over a year and the problem has not been solved, the solution @brminnick does not sound very good to me due to all the efforts made, but it solved this problem in a demo I made, we have an application with 5 million downloads made in xamarin and we often have a crash and only now I discovered that this was the real problem, what a shame.
That is unfortunate 👎

**Comment by niccolocapitelli** [comment]
I had this problem on a production app. I need this fix in the 5.0 milestone.

**Comment by AlleSchonWeg** [comment]
I replaced the CollectionView with a StackLayout and Datatemplate. This works for me.

**Comment by JMNewsome** [comment]
@AlleSchonWeg Replacing a CollectionView with a StackLayout and Datatemplate is not a proper solution as it does not support virtualization, so you will see an impact the more items you have

**Comment by rudyspano** [comment]
Hello. We achieved step by step to deliver a good experience to our customers in terms of performances with our CollectionView containing lot of content (Images, Buttons, Texts, ...). I have one advice and one trick that could perhaps help some of you...

1) Most of the time, it's not only the fault of the CollectionView, I advice you to create a Debug page to see how much time is needed for the view you define as DataTemplate are rendered if you add 10 => 100 instances in a StackLayout (for example). By separating each part into different files, we could see which element are slow to render (StopWatch before and after InitializeComponent). Off course, it depends on the targeted device and configuration (debug vs release, AOT or not, use of HotReload, ...) but we could optimize our cells rendering time by ~3-4. Some controls are surprisingly slow or unstable in terms of rendering (without binding), we replaced several BorderView by a complte Image to limit the number of controls for example... Moreover, do not forget that controls not visible have also a cost... (see https://xamgirl.com/stop-doing-isvisibletrue-false-to-show-hide-views-in-runtime-in-xamarin-forms/ for solutions)

2) We have several types of cells => several DataTemplate and a DataTemplateSelector. In my opitnion, a big problem with the CollectionView implementation on Android (and also on iOS by the way) is that the Adapter has use a few types of ViewHolders (see https://github.com/xamarin/Xamarin.Forms/blob/d99264507dc1623aecb83daf81aadd2effae62ae/Xamarin.Forms.Platform.Android/CollectionView/ItemViewType.cs ). The consequence is that ItemViewType.TemplatedItem was used for all cells and when a ViewHolder changed from a type of view to another (different Datatemplate), a new view was created... On a long list, with long scroll, views were created indefinitively (bad recycling behavior). In short, I suggest you to override the CollectionViewRenderer and the associated Adapter with a logic that creates a ViewHolder for each type of item. For example: 

     ```
private class GroupableItemsViewAdapterAdvanced<TItemsView, TItemsViewSource> : GroupableItemsViewAdapter<TItemsView, TItemsViewSource>
            where TItemsView : GroupableItemsView
            where TItemsViewSource : IGroupableItemsViewSource
        {
            private Dictionary<object, int> _viewHoldersViewTypesAssociations = new Dictionary<object, int>();
            private CollectionViewAdvancedRenderer _collectionViewAdvancedRenderer;

            public GroupableItemsViewAdapterAdvanced(TItemsView groupableItemsView, CollectionViewAdvancedRenderer collectionViewAdvancedRenderer, Func<View, Context, ItemContentView> createView = null)
                : base(groupableItemsView, createView)
            {
                _collectionViewAdvancedRenderer = collectionViewAdvancedRenderer;
            }

            public override int GetItemViewType(int position)
            {
                var itemViewType = base.GetItemViewType(position);

                //ensure all item types have its own placeholder and not a generic one TemplatedItem
                if (itemViewType == ItemViewType.TemplatedItem)
                {
                    var itemType = ItemsSource.GetItem(position).GetType();

                    if (!_viewHoldersViewTypesAssociations.ContainsKey(itemType))
                    {
                        int viewTypeId;

                        if (!_viewHoldersViewTypesAssociations.Any())
                        {
                            //Framework TemplateItem View type is 42, ours are 421,422, ... 
                            viewTypeId = ItemViewType.TemplatedItem * 10 + 1;
                        }
                        else
                        {
                            viewTypeId = _viewHoldersViewTypesAssociations.Max(assocation => assocation.Value) + 1;
                        }

                        //By default, Cells are not always recycled depending on state. Force to recycle the most mossible
                        //https://developer.android.com/reference/android/support/v7/widget/RecyclerView.RecycledViewPool.html#setMaxRecycledViews(int,%20int)
                        _collectionViewAdvancedRenderer.GetRecycledViewPool().SetMaxRecycledViews(viewTypeId, 200);

                        _viewHoldersViewTypesAssociations.Add(itemType, viewTypeId);
                    }

                    itemViewType = _viewHoldersViewTypesAssociations[itemType];
                }

                return itemViewType;
            }
        }
```

**Comment by roubachof** [comment]
I also noticed that the GetItemViewType was not correctly implemented on Android.
You should have one ItemViewType by DataTemplate.
It means that if you have a DataTemplateSelector with 4 differents DataTemplate, you should have one item type for each of those DataTemplate.
See my implementation of the GetItemViewType in the sharpnado's HorizontalListView:

```csharp
            public override int GetItemViewType(int position)
            {
                if (_isDisposed)
                {
                    return -1;
                }

                if (_element.ItemTemplate is DataTemplateSelector dataTemplateSelector)
                {
                    var dataTemplate = dataTemplateSelector.SelectTemplate(_dataSource[position], _element);

                    int itemViewType = _dataTemplates.IndexOf(dataTemplate);
                    if (itemViewType == -1)
                    {
                        itemViewType = _dataTemplates.Count;
                        _dataTemplates.Add(dataTemplate);
                    }

                    return itemViewType;
                }

                return base.GetItemViewType(position);
            }
```

**Comment by angelru** [comment]
The same issue here...

**Comment by azrinsani** [comment]
> I'm seeing the same behavior in my GitTrends app.
> 
> Here's a video one of the users posted on Twitter running GitTrends v1.2.0:
> https://twitter.com/anaseeen/status/1281610967336595458?s=20
> 
> * 4GB RAM
> * Media Tek Helio octa core P23
> * Android 9
> 
> GitTrends v1.2.0 Source Code:
> https://github.com/brminnick/GitTrends/releases/tag/v1.2.0

Wow that's jittery as hell.. should throw away the app or remake with flutter or something... Did you manage to solve this problem?

**Comment by azrinsani** [comment]
I found that correcting the GC solved my collection view scroll jittery problem. Android default is 512KB, which is very small. I changed it to 32MB and its silky smooth now. 

Side Note:
- I use XAML, not code behind.
- I have hella lot of bindings (and mostly two way bindings)
- I use datatemplates, NOT datatemplateselector as advised by Brandon. My item source is huge (like 3000+), but it's silky smooth.
- I use a lot of custom renderers

All these from my playing around, does not seem to effect the scroll smoothness.

**Comment by azrinsani** [comment]
Also to note today I spent a few hours trying to further improve the scrolling, which came back again! I'd like to share what I did that worked

**WHAT DID _NOT_ WORK**
- I **tried swapping my 'Editor's with 'Label's**, as according to your analysis, labels are faster. Unfortunately this had no difference on scroll choppiness. It was still very choppy.
- I **tried reducing the CollectionView itemsource size** (I had 3000 items, i reduced it to 10), though this had no effect on performance at all. Maybe you want to update your info to mention collection size does not effect
- I **tried changing FlexLayout to StackLayout** this too had no impact on performance
- I tried Removing All icon images, this too had not impact on performance
-  I **tried Changing my 'BindableLayout.ItemTemplateSelector'  to only 'BindableLayout.ItemTemplate'**. This does not seem to improve performance as how some have alluded too.
- Some of my controls had DataTrigger. **Removing Data Triggers** did not have any impact on performance
- I had a **verbose alternative declaration of a button** using labels and frames. Changing this to a button did not seem to have any impact on scroll performance.


**WHAT WORKED**
- Removing Animations. I had a popup message which would appear by means of opacity animation. I noticed by removing this improved performance.
- I had a Viewmodel that when 'Get' was called, it perform a few calculations. Removing this improved performance very slightly, though not that significant.
- **OK HERE'S THE BIG ONE** => I **tried removing BindableLayout.ItemsSource** on a flex layout which I was using, and manually added items one by one into the flex layout. Surprise! This had significant jitter improvement. As of now, I know now that this is the main culprit. Sadly, this means I cannot use my view models anymore. On this note, I tried to wrap the FlexLayout inside a <ContentView>, as some suggested though this did not work as well. **I also changed the binding of itemsource to ONETIME binding, also did not work!!!**, this got me thinking that BindableLayout is just bad.

**VERDICT => BINDABLE LAYOUT PERFORMANCE IS BAD.... DONT USE IT** . On further reasearch I realized that this has been mentioned on Microsoft Website:-

_Bindable layouts should only be used when the collection of items to be displayed is small, and scrolling and selection isn't required. While scrolling can be provided by wrapping a bindable layout in a ScrollView, this is not recommended as **bindable layouts lack UI virtualization**. When scrolling is required, a scrollable view that includes UI virtualization, such as ListView or CollectionView, should be used. Failure to observe this recommendation can lead to performance issues._

Source => [https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/layouts/bindable-layouts](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/layouts/bindable-layouts)

**Comment by BlueRaja** [comment]
This is _still_ an issue, and none of the workarounds seem to work.  In my scenario, each item is a single label with a single one-time binding on `Text` _(or 0 bindings, using @brminnick's hack)_.  There's 1000's of them, grouped into groups of 10 using `IsGrouped="True"`.

The first 500 or so items scroll fine, but after that it gets slower and slower until around the 1000th item it's literally unusable.  Somewhere around 1500 items the app crashes.

1000 items should be a fairly common scenario, so it's absurd that there's seemingly no way to get this to work.

**Comment by AlleSchonWeg** [comment]
You could try Sharpnado ListView: https://github.com/roubachof/Sharpnado.HorizontalListView. It supports grid, horizontal and vertical layout.

**Comment by BlueRaja** [comment]
@AlleSchonWeg Yes I'm aware of that one, along with [several other open-source options](https://github.com/xamarin/Xamarin.Forms/issues/8640#issuecomment-839884028) but [none of them support groupings](https://github.com/roubachof/Sharpnado.HorizontalListView/issues/6).

**Comment by Phenek** [comment]
It's really hard to have an Android app that is fluid with Xamarin.
It seems impossible to me

**Comment by othmanTeffahi99** [comment]
I tested the CollectionView on MAUI in Android, it's working great, you can migrate to Maui, but it is still in the preview Version .

**Comment by dimonovdd** [comment]
> `MONO_GC_PARAMS=nursery-size=64m`

@brminnick Hi. Could it create serious problems or memory leak?

**Comment by haavamoa** [comment]
This seems interesting:
https://github.com/levitali/CompiledBindings/issues/4
